### PR TITLE
[FEATURE] array support for the replace() expression function

### PR DESCRIPTION
**Pull request by nirvn**
Simple replace() expression function upgrade making use of the (extremely cool) recent array support addition.

This new string function allows you to modify a string by replacing matching array of strings with either an array of strings or a single string replacement. It can clean up expression that relies in a cascade of replace() calls (i.e. replace(replace(replace(‘boring’)))) .

For e.g., a simple expression to convert numbers from latin to Khmer: 

`replace(my_number,array(‘1’,‘2’,‘3’,‘4’,‘5’,‘6’,‘7’,‘8’,‘9’,’0’),array(‘១’,‘២’,‘៣’,‘៤’,‘៥’,‘៦’,‘៧’,‘៨’,‘៩’,’០’))`

Going from 10 replace() to a single replace() is a nice win.

### CODE REVIEW

**Comment by nirvn** [comment]
@nyalldawson , your review would be appreciated/needed.

File: tests/src/core/testqgsexpression.cpp [hunk 1]
```diff
@@ -821,6 +821,8 @@ class TestQgsExpression: public QObject
       QTest::newRow( "upper" ) << "upper('HeLLo')" << false << QVariant( "HELLO" );
       QTest::newRow( "length" ) << "length('HeLLo')" << false << QVariant( 5 );
       QTest::newRow( "replace" ) << "replace('HeLLo', 'LL', 'xx')" << false << QVariant( "Hexxo" );
+      QTest::newRow( "replace_array (array replaced by array)" ) << "replace_array('321', array('1','2','3'), array('7','8','9'))" << false << QVariant( "987" );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp]: Can you add tests with unbalanced array lengths? Ie one with before > after, and vice versa.

**Review by nyalldawson — CHANGES_REQUESTED** [review]
What do you think about instead modifying the existing replace method to detect array arguments? I'd suggest this is a nicer approach. You could even allow an array for the second argument and a single value for the third to replace any of the listed values with the single replacement.

**Comment by nirvn** [comment]
@nyalldawson 

- [x] I've upgraded replace() instead of creating a new function, glad you suggested that
- [x] the additional invalid tests have been added

Thanks for the quick review.

File: resources/function_help/json/replace [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
-  "description": "Returns a string with the the supplied string replaced.",
+  "description": "Returns a string with the the supplied string or array of strings replaced by a string or an array of strings.",
   "arguments": [ {"arg":"string","description":"the input string"},
-                 {"arg":"before","description":"the string to replace"},
-                 {"arg":"after","description":"the string to use as a replacement"}],
-  "examples": [ { "expression":"replace('QGIS SHOULD ROCK','SHOULD','DOES')", "returns":"'QGIS DOES ROCK'"}
+                 {"arg":"before","description":"the string or array of strings to replace"},
+                 {"arg":"after","description":"the string or array of strings to use as a replacement"}],
+  "examples": [ { "expression":"replace('QGIS SHOULD ROCK','SHOULD','DOES')", "returns":"'QGIS DOES ROCK'"},
+                { "expression":"replace_array('QGIS ABC',array('A','B','C'),array('X','Y','Z'))", "returns":"'QGIS XYZ'"},
```
Comment (DelazJ) [inline resources/function_help/json/replace]: replace_array left here and below

File: resources/function_help/json/replace [hunk 3]
```diff
@@ -1,10 +1,12 @@
 {
   "name": "replace",
   "type": "function",
-  "description": "Returns a string with the the supplied string replaced.",
+  "description": "Returns a string with the the supplied string or array of strings replaced by a string or an array of strings.",
```
Comment (DelazJ) [inline resources/function_help/json/replace]: typo: the the

**Comment by m-kuhn** [comment]
Wasn't map support added as well? I think that would make it even easier to read and maintain such lists.

**Comment by nirvn** [comment]
@DelazJ , thanks for spotting those mistake, the updated commit fixed those.

@m-kuhn , I don't think map would have worked here, especially not now that we've gone down the road of upgrading the replace() function. Relying on a map object would have made it impossible to do replace an array of strings by a single string, i.e. replace('my string','array('y','n','g','etc'),'').

**Comment by m-kuhn** [comment]
@nirvn I was mostly thinking of

```
replace( my_number,
  map('1', '១',
      '2', '២',
	  '3', '៣',
	  '4', '៤',
	  '5', '៥',
	  '6', '៦',
	  '7', '៧',
	  '8', '៨',
	  '9', '៩',
	  '0', '០')
)
```

I think they can live side-by-side?

**Comment by nirvn** [comment]
@m-kuhn , something like what I've just committed to this PR? :)

**Comment by m-kuhn** [comment]
Good job!

File: src/core/qgsexpression.cpp [hunk 4]
```diff
@@ -1214,10 +1214,62 @@ static QVariant fcnLength( const QVariantList& values, const QgsExpressionContex
...
+    for ( int i = 0; i < before.length(); i++ )
+    {
+      str = str.replace( before.at( i ).toString(), after.at( isSingleReplacement ? 0 : i ).toString() );
+    }
+
+    return QVariant( str );
+  }
+  else
+  {
+    parent->setEvalErrorString( QObject::tr( "Function make_point requires 2 or 3 arguments" ) );
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: make_point -> replace

File: resources/function_help/json/replace [hunk 5]
```diff
@@ -1,10 +1,13 @@
 {
   "name": "replace",
   "type": "function",
-  "description": "Returns a string with the the supplied string replaced.",
+  "description": "Returns a string with the supplied string or array of strings replaced by a string or an array of strings. The function can also work with two arguments, whereas the second argument is a map value used to replace keys with associated string values.",
```
Comment (nyalldawson) [inline resources/function_help/json/replace]: Have a look at the "length" function help - that's a good example how variants such be handled for functions.

**Comment by nyalldawson** [comment]
Very nice!

**Comment by nirvn** [comment]
@nyalldawson , done.

**Comment by nyalldawson** [comment]
Fantastic stuff!

**Comment by nirvn** [comment]
@nyalldawson , @m-kuhn , thanks for the review / improvement suggestions.

**Pull request merged**